File: spare_parts_shop_flutter/lib/models/purchase.dart
import 'supplier.dart';
import 'purchase_item.dart';

class Purchase {
  final int id;
  final String invoiceNumber;
  final Supplier supplier;
  final double subtotal;
  final double gstAmount;
  final double discount;
  final double finalAmount;
  final String gstType;
  final String purchaseDate;
  final String? attachmentPath;
  final List<PurchaseItem> items;

  Purchase({
    required this.id,
    required this.invoiceNumber,
    required this.supplier,
    required this.subtotal,
    required this.gstAmount,
    required this.discount,
    required this.finalAmount,
    required this.gstType,
    required this.purchaseDate,
    this.attachmentPath,
    required this.items,
  });

  factory Purchase.fromJson(Map<String, dynamic> json) {
    var itemsList = json['items'] as List;
    List<PurchaseItem> items = itemsList.map((i) => PurchaseItem.fromJson(i)).toList();

    return Purchase(
      id: json['id'],
      invoiceNumber: json['invoiceNumber'],
      supplier: Supplier.fromJson(json['supplier']),
      subtotal: (json['subtotal'] as num).toDouble(),
      gstAmount: (json['gstAmount'] as num).toDouble(),
      discount: (json['discount'] as num).toDouble(),
      finalAmount: (json['finalAmount'] as num).toDouble(),
      gstType: json['gstType'],
      purchaseDate: json['purchaseDate'],
      attachmentPath: json['attachmentPath'],
      items: items,
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'invoiceNumber': invoiceNumber,
      'supplier': supplier.toJson(),
      'subtotal': subtotal,
      'gstAmount': gstAmount,
      'discount': discount,
      'finalAmount': finalAmount,
      'gstType': gstType,
      'purchaseDate': purchaseDate,
      'attachmentPath': attachmentPath,
      'items': items.map((i) => i.toJson()).toList(),
    };
  }
}

File: spare-parts-shop/src/main/java/com/spareparts/model/Purchase.java
package com.spareparts.model;

import com.fasterxml.jackson.annotation.JsonIgnore;
import jakarta.persistence.*;
import lombok.AllArgsConstructor;
import lombok.Data;
import lombok.NoArgsConstructor;
import java.time.LocalDateTime;
import java.util.ArrayList;
import java.util.List;

@Entity
@Table(name = "purchases")
@Data
@NoArgsConstructor
@AllArgsConstructor
public class Purchase {
    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;
    
    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "business_id", nullable = false)
    @JsonIgnore
    private Business business;
    
    @Column(name = "invoice_number", nullable = false)
    private String invoiceNumber;
    
    @ManyToOne(fetch = FetchType.EAGER)
    @JoinColumn(name = "supplier_id", nullable = false)
    private Supplier supplier;
    
    @Column(nullable = false)
    private Double subtotal;
    
    @Column(name = "gst_amount")
    private Double gstAmount = 0.0;
    
    @Column(nullable = false)
    private Double discount = 0.0;
    
    @Column(name = "final_amount", nullable = false)
    private Double finalAmount;
    
    @Column(name = "gst_type")
    private String gstType;
    
    @Column(name = "purchase_date")
    private LocalDateTime purchaseDate = LocalDateTime.now();
    
    @Column(name = "attachment_path")
    private String attachmentPath;
    
    @OneToMany(mappedBy = "purchase", cascade = CascadeType.ALL, orphanRemoval = true)
    private List<PurchaseItem> items = new ArrayList<>();
}
